### PR TITLE
Teach ScaffoldState to hide and show FAB

### PR DESCRIPTION
**Pull request by dnfield**
This is to support #497 

In particular, if you have a FAB with a standard bottom sheet, it should be able to hide like here: https://storage.googleapis.com/spec-host-backup/mio-design%2Fassets%2F10oN2rj2F4Ju3jJUg6wFCRqA_G5TsWQUm%2Fbehavior-visibility-standard.mp4

This also might be useful for other scenarios, like people who want to do what LinkedIn does with their FAB on scrolling.

Broken out to keep the bottom sheet PR from being unnecessarily large.

### CODE REVIEW

**Comment by dnfield** [comment]
I have to fix tests.

**Review by goderbauer — COMMENTED** [review]
Can you add tests?

File: packages/flutter/test/material/scaffold_test.dart [hunk 1]
```diff
@@ -147,6 +147,25 @@ void main() {
...
+        onPressed: null,
+        child: const Text('1'),
+      ),
+    )));
+
+    expect(find.text('1').hitTestable(), findsOneWidget);
+    Scaffold.of(one.currentContext).hideFloatingActionButton();
+    await tester.pumpAndSettle();
+    expect(find.text('1').hitTestable(), findsNothing);
+Scaffold.of(one.currentContext).showFloatingActionButton().orCancel;
```
Comment (goderbauer) [inline packages/flutter/test/material/scaffold_test.dart]: nit: indentation

Why the `orCancel`?

File: packages/flutter/test/material/scaffold_test.dart [hunk 2]
```diff
@@ -147,6 +147,25 @@ void main() {
...
 
+  testWidgets('Floating action button show/hide', (WidgetTester tester) async {
+    final GlobalKey one = new GlobalKey(debugLabel: 'one');
+    await tester.pumpWidget(new MaterialApp(home: new Scaffold(
+      floatingActionButton: new FloatingActionButton(
+        key: one,
+        onPressed: null,
+        child: const Text('1'),
+      ),
+    )));
```
Comment (goderbauer) [inline packages/flutter/test/material/scaffold_test.dart]: nit: add trailing commas to ensure that each property shows up on a separate line.

File: packages/flutter/test/material/scaffold_test.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/test/material/scaffold_test.dart]: Doh.  That's not supposed ot be there.  I was having trouble getting the the right test to validate the FAB wasn't shown anymore.. will fix that.

File: packages/flutter/test/material/scaffold_test.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/test/material/scaffold_test.dart]: This is copy pasted from the test above - I assumed it had been done this way for readability.  Should I change it?

Comment (goderbauer) [reply packages/flutter/test/material/scaffold_test.dart]: It's fine to leave it as is then.

**Comment by Hixie** [comment]
Pretty sure we already have code to hide FABs. cc @DaveShuckerow and @HansMuller who worked on this in the past.

**Comment by dnfield** [comment]
I did see code to hide it during the move transition, but it was tied up with that transition tightly and not public.

**Comment by DaveShuckerow** [comment]
The Fab will auto-hide when it is set to null.

**Comment by dnfield** [comment]
Setting it to null isn't really an option if someone that doesn't own it wants to hide it (temporarily).

Immediate use case is that when a bottom sheet goes full screen, the FAB should hide.  It feels really wrong (and maybe not even possible?) for the bottom sheet to try to find the FAB and set it to null (and causing the whole scaffold to be rebuilt?).

This option allows the FAB to be hidden without rebuilding the scaffold or removing someone else's control.

**Comment by dnfield** [comment]
Discussed this with @HansMuller a bit more - going to look at making this more configurable and rope in some of the other hidden animation stuff we do around moving the FAB so that the animations are themable.

**Comment by DaveShuckerow** [comment]
I'm OK with the goal of this change.  I think all that needs to be done is make sure it modifies/re-uses the existing animation logic so that we don't need another source of entrance/exit animations.

**Comment by dnfield** [comment]
I've updated this patch to make better use of existing code for manipulating the FAB. Scaffold state now exposes four new members:

1.  `bool get hasFloatingActionButton`, which a consumer might care to check if their layout/animation logic needs to worry about the FAB existing at all.
2. `AnimationController get floatingActionButtonVisibilityController`, which a consumer can use to manipulate the value of the FAB visibility animation controller directly.  This could lead to bad things (e.g. if a user called `dispose` on the controller for some reason), but that doesn't seem to make it worth trying to wrap the controller in a more restrictive object.  I don't want to return an `Animation<double>` here, because I want users to be able to write/control the state, e.g. coordinate the FAB dismissing as a bottom sheet approaches full screen.
3. `hideFloatingActionButton` - convenience method for just hiding the FAB, regardless of your own animation state.
4. `showFloatingActionButton` - opposite of 3.

@DaveShuckerow - I've left a signfiicant amount of the state logic in `_FloatingActionButtonTransition`, because rewriting that started to pull almost all of the class it into `ScaffoldState`.  That class still handles some complex behaviors for moving the FAB according to the spec, and it fits with the overall goal I have of letting other consumers manipulate the state of the FAB's visibility animation.

@HansMuller - I didn't come up with a nice way to make the specific animation types.  The logic Dave implemented for transitions is much more likely to go wrong if we expose it too easily than to go right.  There would be a bunch of builders required for scenarios that most users probably don't think about and would easily get confused about. I'd be happy to discuss further if you think there's something  else worth trying here.

**Comment by dnfield** [comment]
So for my immediate needs with BottomSheet, I'd need to have it not just be a on/off thing.  I'd need to be able to tell the scaffold I'm about to use up this percentage of the space the FAB would typically need, and it should hide that much.

I think this would still be fine for most cases - if someone wanted to forward/reverse/fling, they could build their own controller to control that value, which is really just the `AnimationController`'s value.

**Comment by dnfield** [comment]
Closing in favor of #21896

**Pull request closed without merging**